src/Interpreters/Access/InterpreterShowGrantsQuery.cpp

```diff
@@ -154,6 +154,7 @@ std::vector<AccessEntityPtr> InterpreterShowGrantsQuery::getEntities() const
         if (!entity)
             continue;
         if ((id == access->getUserID() /* Any user can see his own grants */)
+            || (entity->isTypeOf<Role>() && access->getRolesInfo()->current_roles.contains(id) /* and grants from the granted roles */)
```

Comment:
It seems it's better to allow here not only current roles but all granted roles + all enabled roles (where enabled roles are current roles and roles granted to the current roles). You can check if a role is granted to a user by writing `getUser()->granted_roles.isGranted()` (however `getUser()` can return null if this is the global context).

Reply:
Indeed. I'm a bit confused about the difference between current roles/enabled roles/granted roles/etc

Reply:
That's easy, for example after the following commands

```
CREATE USER U
CREATE ROLE A, B, C, D

GRANT A, B TO U
GRANT C TO A
GRANT D TO C

-- login as U
SET ROLE A
```

now A is current role (after SET ROLE)
A, B are granted roles (they both are granted to U)
A, C, D are enabled roles (A is current and C is granted to A and D is granted to C)

Reply:
Thanks for explanation. That's very helpful.

You mentioned that `getUser()` can return null in the global context. In which case show grants query can be invoked with the global context?

Reply:
I don't know an example of that :)
But still it seems it's better to do that check in the code just in case (the code may be moved somewhere later where that will be possible).

Reply:
No, enabled roles are close to that, but they are based on current roles instead of granted roles, and current roles is a subset of granted roles.

But I think we can start with enabled roles + list of directly granted roles because that will be enough in 99% cases.